arai60_8.md

```diff
@@ -0,0 +1,133 @@
+# Step 1
+- 解答の形式を掴むのが難しかった。他の方のコードを見てちゃんと理解した。
+# Step 2
+
+## ソートを用いる。リストに点数を格納して、新しいスコアが追加されたらいちいちソートする。
+- まず初期状態のリストを降順にソートして、addメソッドが呼びだされるたびに値を追加してまたソートする。k番目をインデクスで指定して取り出す。
+```python
+class KthLargest:
+
+    def __init__(self, k: int, nums: List[int]):
+        self.k = k
+        self.nums = sorted(nums, reverse=True)
+
+
+    def add(self, val: int) -> int:
+        self.nums.append(val)
+        self.nums.sort(reverse=True)
+        return self.nums[self.k - 1]
+```
+
+## k番目より低い点数はいらないから除外する。addメソッドでは末尾のアイテムを取り出せばそれが答え。
+```python
+class KthLargest:
+
+    def __init__(self, k: int, nums: List[int]):
+        self.k = k
+        self.nums = sorted(nums, reverse=True)
+        if len(self.nums) > self.k:
+            self.nums = self.nums[:self.k]
+
+    def add(self, val: int) -> int:
+        self.nums.append(val)
+        self.nums.sort(reverse=True)
+        if len(self.nums) > self.k:
+            self.nums = self.nums[:self.k]
+        return self.nums[-1]
+```
+
+## 優先度付きキュー(ヒープ)を用いる。
+```python
+class KthLargest:
+
+    def __init__(self, k: int, nums: List[int]):
+        self.k = k
+        self.nums = nums
+        heapq.heapify(self.nums)
+        while self.k < len(self.nums):
+            heapq.heappop(self.nums)
+
+    def add(self, val: int) -> int:
+        heapq.heappush(self.nums, val)
+        while self.k < len(self.nums):
+            heapq.heappop(self.nums)
+        return self.nums[0]
+```
+- heapq.heapify()は破壊的。
+
+### heapifyの代わりにheappushを使用
+```python
+class KthLargest:
+    def __init__(self, k: int, nums: List[int]):
+        self.k = k
+        self.top_k_scores = []  #heap
+        for num in nums:
+            heapq.heappush(self.top_k_scores, num)
+        while self.k < len(self.top_k_scores):
+            heapq.heappop(self.top_k_scores)
+
+    def add(self, val: int) -> int:
+        heapq.heappush(self.top_k_scores, val)
+        while self.k < len(self.top_k_scores):
+            heapq.heappop(self.top_k_scores)
+        return self.top_k_scores[0]
+```
+# Step 3
+## リストを用いる方法
+```python
+class KthLargest:
+
+    def __init__(self, k: int, nums: List[int]):
+        self.k = k
+        self.top_k_scores = nums
+        self.top_k_scores.sort(reverse=True)
+        if k < len(self.top_k_scores):
+            self.top_k_scores = self.top_k_scores[:k]
+
+
+    def add(self, val: int) -> int:
+        self.top_k_scores.append(val)
+        self.top_k_scores.sort(reverse=True)
+        if self.k < len(self.top_k_scores):
+            self.top_k_scores = self.top_k_scores[:self.k]
+        return self.top_k_scores[-1]
+```
+## heapを用いる方法
+```python
+class KthLargest:
+
+    def __init__(self, k: int, nums: List[int]):
+        self.k = k
+        self.top_k_scores = []  #heap
```

Comment:
[To improve legibility, these comments should start at least 2 spaces away from the code with the comment character #, followed by at least one space before the text of the comment itself.](https://google.github.io/styleguide/pyguide.html#2144-decision:~:text=to%20improve%20legibility%2C%20these%20comments%20should%20start%20at%20least%202%20spaces%20away%20from%20the%20code%20with%20the%20comment%20character%20%23%2C%20followed%20by%20at%20least%20one%20space%20before%20the%20text%20of%20the%20comment%20itself.)

[An inline comment is a comment on the same line as a statement. Inline comments should be separated by at least two spaces from the statement. They should start with a # and a single space.](https://peps.python.org/pep-0008/#inline-comments:~:text=an%20inline%20comment%20is%20a%20comment%20on%20the%20same%20line%20as%20a%20statement.%20inline%20comments%20should%20be%20separated%20by%20at%20least%20two%20spaces%20from%20the%20statement.%20they%20should%20start%20with%20a%20%23%20and%20a%20single%20space.)